### ISSUE TITLE
Should we implement a Docker Healthcheck ? Yes / No, what should we validate ? 

### ISSUE THREAD

**Issue by JnMik** [Mashape/docker-kong#89]
I'm creating the issue to discuss what good a Docker Healthcheck could bring to the image, and how should we achieve it. 

**Problem I have** (Previously discussed here : https://github.com/Mashape/kong/issues/2656)
Joining a new Kong container in a Kong cluster could cause some HTTP connexion refused if Kong / NGINX is slow to start or if the "Kong start" command is delayed to achieve other tasks before starting.

I believe a Healthcheck in Dockerfile doing some validations would be nice. 
We could call a shell script so we could easily validate multiple options 

Here's some thoughts on the possible options to check : 

- Check if the needed ports are open in netstats or ss. (Based on that we would at least know that kong is running and listening)

- Would also be nice to have some kind of check that make sure that kong has at least pull the initial data from the database. (But maybe kong starts listening on ports after that, so that point would already be covered)

Any other suggestions or are you against the idea of having a Docker Healthcheck ?
Let's discuss this.

Cheers !

**Comment by Tieske** [comment]
Once a connection is being accepted on the proxy port, Kong should be available.

As an http alternative you can create an api for some hostname eg. `healthcheck.com`, and configure it with the `request-termination` plugin returning a custom response.

**Comment by JnMik** [comment]
I used to have that workaround, when I was load balancing using Haproxy or F5.
But I can't use it the same way anymore with the docker swarm service layer load balancing because my first Kong container will never become healthy.

I need the firstly deployed Kong container to be considered healthy inside the swarm cluster before I run Kongfig to configure it, or Kongfig won't be able to reach it through the swarm service layer (internal load balancing). 

So from my understanding, only checking if the port 8000 or 8001 is open could do the trick ?

**Comment by JnMik** [comment]
So here's how I dealt with it

**Dockerfile**
```
FROM kong:0.10.3

# The first kong container will have to run migrations on the database, so it will
# take a while for the port 8000 to be reachable. --start-period will help us delay
# the healthcheck failure
HEALTHCHECK --interval=5s --retries=10 --start-period=180s \
    CMD curl -I -s -L http://127.0.0.1:8000 || exit 1
```

No more dropped connection during Siege benchmark

**Comment by endeepak** [comment]
If you are using docker-compose file to deploy stack in swarm,  you can define heathcheck in compose file as outlined in https://docs.docker.com/compose/compose-file/#healthcheck

**Comment by JnMik** [comment]
@endeepak True ! 
So shall we drop the healthcheck from the Dockerfile, and let the user decide the healthcheck based on his setup and needs ?
